### PR TITLE
Premigrations + Updated NuGet Dependencies

### PR DESCRIPTION
**Pull request by bergmania**
### Description
This PR adds a new concept called _**Premigrations**_.
A premigration is a regular migration, but the premigration plan is always executed unattended and before other migrations.

This have been required to avoid a chicken and the egg paradox, becuase OpenIddict have been updated to a new major. This required some migrations of they EF Core tables that is used to do authentication. So we cannot sign in before the migration is executed, and regular migrations requires a sign-in if not executed unattended.

The following dependencies have been updated
| Name  | Old version | New version |
| ------------- | ------------- | ------------- |
| Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation | 8.0.1  | 8.0.3 |
| Microsoft.Data.Sqlite | 8.0.1  | 8.0.3 |
| Microsoft.EntityFrameworkCore.Sqlite | 8.0.1  | 8.0.3 |
| Microsoft.EntityFrameworkCore.SqlServer | 8.0.1  | 8.0.3 |
| Microsoft.Extensions.FileProviders.Embedded | 8.0.1  | 8.0.3 |
| Microsoft.Extensions.Identity.Core | 8.0.1  | 8.0.3 |
| Microsoft.Extensions.Identity.Stores | 8.0.1  | 8.0.3 |
| Microsoft.Extensions.Options | 8.0.1  | 8.0.3 |
| Asp.Versioning.Mvc| 8.0.0  | 8.1.0 |
| Asp.Versioning.Mvc.ApiExplorer | 8.0.0  | 8.1.0 |
| HtmlAgilityPack | 1.11.57  | 1.11.60 |
| K4os.Compression.LZ4 | 1.3.6 | 1.3.8 |
| MailKit | 4.3.0 | 4.4.0 |
| OpenIddict.Abstractions | 4.10.1 | 5.4.0 |
| OpenIddict.AspNetCore | 4.10.1 | 5.4.0 |
| OpenIddict.EntityFrameworkCore | 4.10.1 | 5.4.0 |
| SixLabors.ImageSharp.Web| 3.1.0 | 3.1.1 |

Beside these that is shipped, we have also updated:

| Name  | Old version | New version | Note |
| ------------- | ------------- | ------------- | ------------- |
| Microsoft.EntityFrameworkCore.Design | 8.0.1  | 8.0.3 | Only used to use dotnet ef command to create migrations |
| NJsonSchema | 10.9.0  | 11.0.0 | Only used in our project to create the schema file |
| Microsoft.NET.Test.Sdk | 17.8.0  | 17.9.0 | Only used in test projects |
| Bogus | 34.0.2  | 35.5.0 | Only used in test projects |
| Microsoft.AspNetCore.Mvc.Testing | 8.0.1  | 8.0.3 | Only used in test projects |

### Test
- [x] Verify you can install a clean install using SqlServer
- [x] Verify you can install a clean install using Sqlite
- [x] Verify you can update from v14/dev database with SqlServer
- [x] Verify you can update from v14/dev database with Sqlite
- [ ] Verify you can update from 13.0.0 database with SqlServer
- [x] Verify you can update from 13.0.0 database with Sqlite
- [x] Smoke test things works as expected

### CODE REVIEW

**Comment by nzdev** [comment]
Perhaps migrations could move to unattended only? 
I would guess that once pre-migrations have run, that older code won't be forward compatible with database changes and new code won't run without updating the schema.

File: src/Umbraco.Infrastructure/Runtime/CoreRuntime.cs [hunk 1]
```diff
@@ -179,6 +179,26 @@ private async Task StartAsync(CancellationToken cancellationToken, bool isRestar
             throw new InvalidOperationException($"An instance of {typeof(IApplicationShutdownRegistry)} could not be resolved from the container, ensure that one if registered in your runtime before calling {nameof(IRuntime)}.{nameof(StartAsync)}");
         }
 
+        var premigrationUpgradeNotification = new RuntimePremigrationsUpgradeNotification();
+        await _eventAggregator.PublishAsync(premigrationUpgradeNotification, cancellationToken);
+        switch (premigrationUpgradeNotification.UpgradeResult)
+        {
+            case RuntimePremigrationsUpgradeNotification.PremigrationUpgradeResult.HasErrors:
+                if (State.BootFailedException == null)
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/Runtime/CoreRuntime.cs:187]:
```suggestion
                if (State.BootFailedException is null)
```

File: src/Umbraco.Infrastructure/Install/PremigrationUpgrader.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            return Task.CompletedTask;
+        }
+
+        using (!_profilingLogger.IsEnabled(LogLevel.Verbose) ? null : _profilingLogger.TraceDuration<UnattendedUpgrader>(
+                   "Starting premigration upgrade.",
+                   "Unattended premigration completed."))
+        {
+            DatabaseBuilder.Result? result = _databaseBuilder.UpgradeSchemaAndData(plan);
+            if (result?.Success == false)
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/Install/PremigrationUpgrader.cs:63]:
```suggestion
            if (result?.Success is false)
```

File: src/Umbraco.Infrastructure/Install/PremigrationUpgrader.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            return Task.CompletedTask;
+        }
+
+        var plan = new UmbracoPremigrationPlan();
+        if (HasMissingPremigrations(plan) is false)
+        {
+            return Task.CompletedTask;
+        }
+
+        using (!_profilingLogger.IsEnabled(LogLevel.Verbose) ? null : _profilingLogger.TraceDuration<UnattendedUpgrader>(
```
Comment (Zeegaan) [inline src/Umbraco.Infrastructure/Install/PremigrationUpgrader.cs:58]: personal preference, don't like `!` 😛 
```suggestion
        using (_profilingLogger.IsEnabled(LogLevel.Verbose) is false ? null : _profilingLogger.TraceDuration<UnattendedUpgrader>(
```

**Review by Zeegaan — CHANGES_REQUESTED** [review]
Found a few small things before testing 😁

**Comment by Zeegaan** [comment]
Seems like there is an error when migrating from v13 -> v14 on SqlServer, but this also happens on `v14/dev´, and has nothing to do with this PR 😁 
Let's get it merged 🚀

**Pull request merged**